client/src/rpc_client.rs

```diff
@@ -25,12 +26,187 @@ use solana_sdk::{
     signature::{KeypairUtil, Signature},
     transaction::{self, Transaction, TransactionError},
 };
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{channel, Receiver};
+use std::sync::{Arc, RwLock};
+use std::thread::JoinHandle;
 use std::{
     error, io,
     net::SocketAddr,
     thread::sleep,
     time::{Duration, Instant},
 };
+use thiserror::Error;
+use tungstenite::{client::AutoStream, connect, Message, WebSocket};
+use url::{ParseError, Url};
+
+#[derive(Debug, Error)]
+pub enum PubsubClientError {
+    #[error("url parse error")]
+    UrlParseError(#[from] ParseError),
+
+    #[error("unable to connect to server")]
+    ConnectionError(#[from] tungstenite::Error),
+
+    #[error("json parse error")]
+    JsonParseError(#[from] serde_json::error::Error),
+
+    #[error("unexpected message format")]
+    UnexpectedMessageError,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct SlotInfoMessage {
+    pub parent: u64,
+    pub root: u64,
+    pub slot: u64,
+}
+
+pub struct PubsubClientSlotSubscription {
```

Comment:
How about making this into a generic PubsubClientSubscription struct so it could be used for other subscription types?
It's not hard to imagine we might use a PubsubClient::account_subscribe down the road.

I think the only substantive changes that need to be made are: passing in the jsonrpc "method" into the `send_subscribe()` and `send_unsubscribe()` methods; and updating `read_message()` to return a generic `Result<T, PubsubClientError> `

Reply:
OK, I think I got it! it was slightly tricky b/c of `T: DeserializeOwned` bound required